adal/src/main/java/com/microsoft/aad/adal/AuthenticationContext.java

```diff
@@ -141,28 +145,77 @@ public AuthenticationContext(Context appContext, String authority, boolean valid
      *                        tokens. Set to null if you don't want cache.
      */
-    public AuthenticationContext(Context appContext, String authority,
-                                 ITokenCacheStore tokenCacheStore) {
+    public AuthenticationContext(@NonNull final Context appContext,
+                                 @NonNull final String authority,
+                                 @Nullable final ITokenCacheStore tokenCacheStore) {
         initialize(appContext, authority, tokenCacheStore, true, false);
     }
 
-    private void initialize(Context appContext, String authority, ITokenCacheStore tokenCacheStore,
-                            boolean validateAuthority, boolean defaultCache) {
+    private void initialize(@NonNull final Context appContext,
+                            @NonNull final String authority,
+                            @Nullable final ITokenCacheStore tokenCacheStore,
+                            boolean validateAuthority,
+                            boolean defaultCache) {
         if (appContext == null) {
             throw new IllegalArgumentException("appContext");
         }
+
         if (authority == null) {
             throw new IllegalArgumentException("authority");
         }
+
         mBrokerProxy = new BrokerProxy(appContext);
+
         if (!defaultCache && !mBrokerProxy.canUseLocalCache(authority)) {
             throw new UnsupportedOperationException("Local cache is not supported for broker usage");
         }
+
         mContext = appContext;
         checkInternetPermission();
         mAuthority = extractAuthority(authority);
         mValidateAuthority = validateAuthority;
-        mTokenCacheStore = tokenCacheStore;
+
+        if (null != tokenCacheStore) {
+            mTokenCacheStore = wrapCache(tokenCacheStore);
+        }
+    }
+
+    private ITokenCacheStore wrapCache(@NonNull final ITokenCacheStore originalCache) {
+        return new ITokenCacheStore() {
+            @Override
+            public TokenCacheItem getItem(final String key) {
+                return originalCache.getItem(key);
+            }
+
+            @Override
+            public Iterator<TokenCacheItem> getAll() {
+                return originalCache.getAll();
+            }
+
+            @Override
+            public boolean contains(final String key) {
+                return originalCache.contains(key);
+            }
+
+            @Override
+            public void setItem(final String key, final TokenCacheItem item) {
+                originalCache.setItem(key, item);
+            }
+
+            @Override
+            public void removeItem(final String key) {
+                originalCache.removeItem(key);
+            }
+
+            @Override
+            public void removeAll() {
+                // Clear our original cache
+                originalCache.removeAll();
+
+                // clear our replica cache
+                getMsalOAuth2TokenCache(mContext).clearAll();
+            }
```

Comment:
do we want to synchronize here, and, indeed, for all of these?

Reply:
Like so?
https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/1552/files#r469607985

Reply:
That's the thought.  Not having setItem have undefined behavior with respect to removeAll.

Reply:
Applied in 7a26dbee681f6ab9fd5fe8a7d10ff13d586414fe